src/txmempool.cpp

```diff
@@ -99,96 +139,338 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
             if (fRecursive) {
-                for (unsigned int i = 0; i < tx.vout.size(); i++) {
+                for (unsigned int i = 0; i < tx.GetVout().size(); i++) {
                     std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(COutPoint(hash, i));
                     if (it == mapNextTx.end())
                         continue;
-                    txToRemove.push_back(it->second.ptx->GetHash());
+                    objToRemove.push_back(it->second.ptx->GetHash());
+                }
+                for(const auto& sc: tx.GetVscCcOut()) {
+                    if (mapSidechains.count(sc.scId) == 0)
+                        continue;
+
+                    if (removeDependantFwds) {
+                        for(const auto& ccObjHash : mapSidechains.at(sc.scId).fwdTransfersSet)
+                            objToRemove.push_back(ccObjHash);
+                    } else
+                        objToRemove.push_back(mapSidechains.at(sc.scId).scCreationTxHash);
+
+                    //no backward cert for unconfirmed sidechain can be in mempool
+                    assert(mapSidechains.at(sc.scId).backwardCertificate.IsNull());
                 }
             }
-            BOOST_FOREACH(const CTxIn& txin, tx.vin)
+
+            BOOST_FOREACH(const CTxIn& txin, tx.GetVin())
                 mapNextTx.erase(txin.prevout);
-            BOOST_FOREACH(const JSDescription& joinsplit, tx.vjoinsplit) {
+            BOOST_FOREACH(const JSDescription& joinsplit, tx.GetVjoinsplit()) {
                 BOOST_FOREACH(const uint256& nf, joinsplit.nullifiers) {
                     mapNullifiers.erase(nf);
                 }
             }
 
-            removed.push_back(tx);
+            for(const auto& fwd: tx.GetVftCcOut()) {
+                if (mapSidechains.count(fwd.scId)) { //Guard against double-delete on multiple fwds toward the same sc in same tx
+                    mapSidechains.at(fwd.scId).fwdTransfersSet.erase(tx.GetHash());
+
+                    if (mapSidechains.at(fwd.scId).fwdTransfersSet.size() == 0 &&
+                        mapSidechains.at(fwd.scId).scCreationTxHash.IsNull() &&
+                        mapSidechains.at(fwd.scId).backwardCertificate.IsNull() )
+                    {
+                        LogPrint("mempool", "%s():%d - erasing [%s] from mapSidechain\n", __func__, __LINE__, fwd.scId.ToString() );
+                        mapSidechains.erase(fwd.scId);
+                    }
+                }
+            }
+
+            for(const auto& sc: tx.GetVscCcOut()) {
+                assert(mapSidechains.count(sc.scId) != 0);
+                mapSidechains.at(sc.scId).scCreationTxHash.SetNull();
+
+                if (mapSidechains.at(sc.scId).fwdTransfersSet.size() == 0)
+                {
+                    LogPrint("mempool", "%s():%d - erasing [%s] from mapSidechain\n", __func__, __LINE__, sc.scId.ToString() );
+                    mapSidechains.erase(sc.scId);
+                }
```

Comment:
if `fwdTransfersSet` is not empty, I guess we should add that transactions to `objToRemove` list.
In other case we may left Transaction with FT to some SC, that was not created and hash no TX with creation output in the mempool.

Reply:
I disagree. This section of the code run for both fRecursive and non fRecursive scenario.
Consider the scenario where you have in mempool both scCreation and fwdTx, and then the scCreation only in included in the block. You should remove only the scCreation from mempool while the fwds should be kept int

Reply:
Agree

Reply:
Done